traffic_ops/traffic_ops_golang/role/roles.go

```diff
@@ -273,3 +317,502 @@ RETURNING id, last_updated`
+			return
+		} else if !ok {
+			api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("no such role"), nil)
+			return
+		}
+
+		existingLastUpdated, found, err := api.GetLastUpdated(inf.Tx, roleID, "role")
+		if err == nil && found == false {
+			api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("no role found with that ID"), nil)
+			return
+		}
+		if err != nil {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+			return
+		}
+
+		if !api.IsUnmodified(r.Header, *existingLastUpdated) {
+			api.HandleErr(w, r, tx, http.StatusPreconditionFailed, api.ResourceModifiedError, nil)
+			return
+		}
+		rows, err := tx.Query(updateRoleQuery(), roleDesc, roleName)
+		if err != nil {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("updating role: "+err.Error()))
+			return
+		}
+		defer rows.Close()
+		if !rows.Next() {
+			api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("no role found with this ID"), nil)
+			return
+		}
+		lastUpdated := tc.TimeNoMod{}
+		for rows.Next() {
+			if err := rows.Scan(&lastUpdated); err != nil {
+				api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("scanning lastUpdated from role update: "+err.Error()))
+				return
+			}
+			roleV4.LastUpdated = &lastUpdated.Time
+		}
+	} else {
+		if err := json.NewDecoder(r.Body).Decode(&role); err != nil {
+			api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+			return
+		}
+		if err := role.validate(inf.Tx); err != nil {
+			api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("name and/or description and/or privLevel can not be empty"), nil)
+			return
+		}
+		roleName = *role.Name
+		roleDesc = *role.Description
+		privLevel = *role.PrivLevel
+		roleCapabilities = role.Capabilities
+		if roleIDParam, ok := inf.Params["id"]; !ok {
+			api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("must supply a role ID to update"), nil)
+			return
+		} else {
+			roleID, err = strconv.Atoi(roleIDParam)
+			if err != nil {
+				api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("must supply an integral role ID to update"), nil)
+				return
+			}
+		}
+		if privLevel > inf.User.PrivLevel {
+			api.HandleErr(w, r, tx, http.StatusForbidden, errors.New("can not create a role with a higher priv level than your own"), nil)
+			return
+		}
+		existingLastUpdated, found, err := api.GetLastUpdated(inf.Tx, roleID, "role")
+		if err == nil && found == false {
+			api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("no role found with that ID"), nil)
+			return
+		}
+		if err != nil {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+			return
+		}
+
+		if !api.IsUnmodified(r.Header, *existingLastUpdated) {
+			api.HandleErr(w, r, tx, http.StatusPreconditionFailed, api.ResourceModifiedError, nil)
+			return
+		}
+
+		rows, err := tx.Query(updateLegacyRoleQuery(), roleName, roleDesc, roleID)
```

Comment:
If we aren't reading the rows, this can just be an `Exec` - but I think we should be reading back `lastUpdated` at least.